internal/rbd/clone.go

```diff
@@ -0,0 +1,236 @@
+		if err == nil {
+			return true, nil
+		}
+		return false, err
+	}
+	if !errors.As(err, &einf) {
+		// return error if its not image not found
+		return false, err
+	}
+
+	err = tempClone.checkSnapExists(snap)
+	if err != nil {
+		if errors.As(err, &esnf) {
+			// check temporary image needs flatten, if yes add task to flatten the
+			// temporary clone
+			err = tempClone.flattenRbdImage(ctx, rv.conn.Creds, false, rbdHardMaxCloneDepth, rbdSoftMaxCloneDepth)
+			if err != nil {
+				return false, err
+			}
+			// as the snapshot is not present, create new snapshot,clone and
+			// delete the temporary snapshot
+			err = createRBDClone(ctx, tempClone, rv, snap, rv.conn.Creds)
+			if err != nil {
+				return false, err
+			}
+			return true, nil
+		} else if !errors.As(err, &einf) {
+			// any error other than image not found return error
+			return false, err
+		}
+	} else {
+		// snap will be create after we flatten the temporary cloned image,no
+		// need to check for flatten here.
+		// as the snap exists,create clone image and delete temporary snapshot
+		// and add task to flatten temporary cloned image
+		err = rv.cloneRbdImageFromSnapshot(ctx, snap)
+		if err != nil {
+			klog.Errorf(util.Log(ctx, "failed to clone rbd image %s from snapshot %s: %v"), rv.RbdImageName, snap.RbdSnapName, err)
+			err = fmt.Errorf("failed to clone rbd image %s from snapshot %s: %v", rv.RbdImageName, snap.RbdSnapName, err)
+			return false, err
+		}
+		err = tempClone.deleteSnapshot(ctx, snap)
+		if err != nil {
+			klog.Errorf(util.Log(ctx, "failed to delete snapshot: %v"), err)
+			return false, err
+		}
+		return true, nil
+	}
+	// as the temp clone doesnot exists,check snapshot exists on parent volume
+	err = parentVol.checkSnapExists(snap)
+	if err == nil {
+		// the temp clone exists, delete it lets reserve a new ID and
+		// create new resources for clearner approach
+		err = parentVol.deleteSnapshot(ctx, snap)
+	}
+	if errors.As(err, &esnf) {
+		return false, nil
+	}
+	return false, err
+}
+
+func (rv *rbdVolume) generateTempClone() *rbdVolume {
+	tempClone := rbdVolume{}
+	tempClone.conn = rv.conn
+	// The temp clone image need to have deep flatten feature
+	f := []string{librbd.FeatureNameLayering, librbd.FeatureNameDeepFlatten}
+	tempClone.imageFeatureSet = librbd.FeatureSetFromNames(f)
+	tempClone.ClusterID = rv.ClusterID
+	tempClone.Monitors = rv.Monitors
+	tempClone.Pool = rv.Pool
+	// The temp cloned image name will be always (rbd image name + "-temp")
+	// this name will be always unique, as cephcsi never creates an image with
+	// this format for new rbd images
+	tempClone.RbdImageName = rv.RbdImageName + "-temp"
+	return &tempClone
+}
+
+func (rv *rbdVolume) createCloneFromImage(ctx context.Context, parentVol *rbdVolume) error {
+	// generate temp cloned volume
+	tempClone := rv.generateTempClone()
+	snap := &rbdSnapshot{
+		RbdSnapName: rv.RbdImageName,
+		Pool:        rv.Pool,
+	}
+
+	var (
+		iErr error
```

Comment:
Please comment these errors or give them proper names, I am wondering what `i` and `f` stand for. While reading code, nobody should need to think twice about these minor things.

`iErr` might better be called `errFlatten`? (this took me surely 2 minutes to figure out, that could have been saved when proper naming/comments were used)